doc/source/reading.rst

```diff
@@ -0,0 +1,476 @@
+
+In the previous tutorials we have provided a fair amount of examples on
+reading/slicing dense and sparse arrays. Specifically, we explained how
+to set your buffers to a read query in order to retrieve the results. We also
+showed that when reading sparse arrays you retrieve only results for the
+non-empty cells, whereas when reading dense arrays TileDB returns special
+fill values for any empty/non-populated cells. Finally, we demonstrated
+how to read only a subset of the array attributes, and also that you can
+even retrieve the coordinates of the result cells in the case of dense arrays.
+
+In this tutorial we will cover a few more important topics on reading,
+such as reading in different layouts, determining the memory you need
+to allocate to your buffers that will hold the results, how to retrieve
+the non-empty domain of your array, and the concept of "incomplete" queries.
+
+.. toggle-header::
+    :header: **Example Code Listing #1**
+
+    .. content-tabs::
+
+       .. tab-container:: cpp
+          :title: C++
+
+          .. literalinclude:: {source_examples_path}/cpp_api/reading_dense_layouts.cc
+             :language: c++
+             :linenos:
+
+.. toggle-header::
+    :header: **Example Code Listing #2**
+
+    .. content-tabs::
+
+       .. tab-container:: cpp
+          :title: C++
+
+          .. literalinclude:: {source_examples_path}/cpp_api/reading_sparse_layouts.cc
+             :language: c++
+             :linenos:
+
+.. toggle-header::
+    :header: **Example Code Listing #3**
+
+    .. content-tabs::
+
+       .. tab-container:: cpp
+          :title: C++
+
+          .. literalinclude:: {source_examples_path}/cpp_api/reading_incomplete.cc
+             :language: c++
+             :linenos:
+
+
+Basic concepts and definitions
+------------------------------
+
+.. toggle-header::
+    :header: **Read query layout**
+
+      You can read from a TileDB array with different layouts (namely row-major,
+      column-major and global order). The row-/column-major layouts are
+      with respect to your defined subarray and may differ from the physical
+      layout of the cells on disk, attributing a lot of flexibility.
+
+.. toggle-header::
+    :header: **Maximum buffer sizes**
+
+      One great challenge when reading sparse arrays or arrays with
+      variable-length attributes (dense or sparse) is allocating the result
+      buffers to be set to the query. The reason is that
+      it is not possible to know the exact result size and, hence,
+      it is very hard to calculate how much memory you need to allocate
+      to your buffers. TileDB provides a function for computing a
+      good upper bound estimate on the buffer size that can certainly
+      hold the result for a given attribute.
+
+
+.. toggle-header::
+    :header: **Non-empty domain**
+
+      TileDB provides a useful function for determining the non-empty
+      domain of your array, i.e., a tight hyper-rectangle that encompasses
```

Comment:
Possibly important to emphasize here that it's not "a tight hyperrectangle" but "the tightest hyperrectangle".